sdk/go/common/util/yamlutil/edit.go

```diff
@@ -26,105 +26,129 @@ type HasRawValue interface {
 	if err != nil {
 		return nil, err
 	}
 
-	newValue, err = editNodes(&oldDoc, &newValue)
+	newNode, err = editNodes(&oldDoc, &newNode)
 	if err != nil {
 		return nil, err
 	}
 
-	return YamlEncode(&newValue)
+	return YamlEncode(&newNode)
 }
 
-func editNodes(original, new *yaml.Node) (yaml.Node, error) {
-	if original.Kind != new.Kind {
-		return *new, nil
+func editNodes(original, newNode *yaml.Node) (yaml.Node, error) {
+	if original.Kind != newNode.Kind {
+		return *newNode, nil
 	}
 
 	ret := *original
-	ret.Tag = new.Tag
-	ret.Value = new.Value
+	ret.Tag = newNode.Tag
+	ret.Value = newNode.Value
 
 	switch original.Kind {
 	case yaml.DocumentNode, yaml.SequenceNode:
 		var minLen int
 		var content []*yaml.Node
-		if len(new.Content) < len(original.Content) {
-			minLen = len(new.Content)
-		} else {
-			minLen = len(original.Content)
-		}
+		minLen = min(len(newNode.Content), len(original.Content))
 
 		for i := 0; i < minLen; i++ {
-			item, err := editNodes(original.Content[i], new.Content[i])
+			item, err := editNodes(original.Content[i], newNode.Content[i])
 			if err != nil {
 				return ret, err
 			}
 			content = append(content, &item)
 		}
 		// Any excess nodes in the new value are copied verbatim
-		content = append(content, new.Content[minLen:]...)
+		content = append(content, newNode.Content[minLen:]...)
 
 		ret.Content = content
 		return ret, nil
 	case yaml.MappingNode:
-		origKeys := make(map[string]int)
-		newKeys := make(map[string]int)
-		var newKeyList []string
-
-		var content []*yaml.Node
-		for i := 0; i < len(original.Content); i += 2 {
-			origKeys[original.Content[i].Value] = i
-		}
-		for i := 0; i < len(new.Content); i += 2 {
-			value := new.Content[i].Value
-			newKeys[value] = i
-			newKeyList = append(newKeyList, value)
-		}
-		for _, k := range newKeyList {
-			newIdx := newKeys[k]
-			origIdx, has := origKeys[k]
-			var err error
-			var key yaml.Node
-			var value yaml.Node
-			if has {
-				key, err = editNodes(original.Content[origIdx], new.Content[newIdx])
-				if err != nil {
-					return ret, err
-				}
-				value, err = editNodes(original.Content[origIdx+1], new.Content[newIdx+1])
-				if err != nil {
-					return ret, err
-				}
-			} else {
-				key = *new.Content[newIdx]
-				value = *new.Content[newIdx+1]
-			}
-			content = append(content, &key)
-			content = append(content, &value)
+		content, err := handleMappingNode(original, newNode)
+		if err != nil {
+			return ret, err
 		}
 
 		ret.Content = content
 		return ret, nil
 	case yaml.ScalarNode, yaml.AliasNode:
-		ret.Content = new.Content
+		ret.Content = newNode.Content
 		return ret, nil
 	}
 	return yaml.Node{}, errors.New("unknown node type")
 }
+
+func handleMappingNode(original, newNode *yaml.Node) ([]*yaml.Node, error) {
+	origKeys := make(map[string]int)
+	newKeys := make(map[string]int)
+	var origKeyList, newKeyList []string
+
+	var content []*yaml.Node
+	for i := 0; i < len(original.Content); i += 2 {
```

Comment:
I see this is also the case in the original code, but I'm a bit confused why we need `i += 2` here?  Would be great to have a comment here for that.  I'm guessing it's because content is a list of key/value "pairs"?

Edit: reading below that's indeed the case, but I still think a comment here would be useful.

Reply:
I agree i also was like wot.  I'll make it clear :)